File: android_webview/native/aw_contents_statics.cc
// Copyright 2014 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "android_webview/native/aw_contents_statics.h"

#include "android_webview/browser/aw_browser_context.h"
#include "android_webview/browser/net/aw_url_request_context_getter.h"
#include "android_webview/common/aw_version_info_values.h"
#include "android_webview/native/aw_contents_io_thread_client_impl.h"
#include "base/android/jni_string.h"
#include "base/android/scoped_java_ref.h"
#include "base/callback.h"
#include "content/public/browser/browser_thread.h"
#include "content/public/browser/storage_partition.h"
#include "content/public/common/url_constants.h"
#include "jni/AwContentsStatics_jni.h"
#include "net/cert/cert_database.h"

using base::android::AttachCurrentThread;
using base::android::ConvertJavaStringToUTF8;
using base::android::JavaParamRef;
using base::android::JavaRef;
using base::android::ScopedJavaGlobalRef;
using base::android::ScopedJavaLocalRef;
using content::BrowserThread;

namespace android_webview {

namespace {

void ClientCertificatesCleared(const JavaRef<jobject>& callback) {
  DCHECK_CURRENTLY_ON(BrowserThread::UI);
  JNIEnv* env = AttachCurrentThread();
  Java_AwContentsStatics_clientCertificatesCleared(env, callback);
}

void NotifyClientCertificatesChanged() {
  DCHECK_CURRENTLY_ON(BrowserThread::IO);
  net::CertDatabase::GetInstance()->OnAndroidKeyStoreChanged();
}

}  // namespace

// static
void ClearClientCertPreferences(JNIEnv* env,
                                const JavaParamRef<jclass>&,
                                const JavaParamRef<jobject>& callback) {
  DCHECK_CURRENTLY_ON(content::BrowserThread::UI);
  BrowserThread::PostTaskAndReply(
      BrowserThread::IO, FROM_HERE,
      base::Bind(&NotifyClientCertificatesChanged),
      base::Bind(&ClientCertificatesCleared,
                 ScopedJavaGlobalRef<jobject>(env, callback)));
}

// static
ScopedJavaLocalRef<jstring> GetUnreachableWebDataUrl(
    JNIEnv* env,
    const JavaParamRef<jclass>&) {
  return base::android::ConvertUTF8ToJavaString(
      env, content::kUnreachableWebDataURL);
}

// static
void SetLegacyCacheRemovalDelayForTest(JNIEnv*,
                                       const JavaParamRef<jclass>&,
                                       jlong delay_ms) {
  AwBrowserContext::SetLegacyCacheRemovalDelayForTest(delay_ms);
}

// static
ScopedJavaLocalRef<jstring> GetProductVersion(JNIEnv* env,
                                              const JavaParamRef<jclass>&) {
  return base::android::ConvertUTF8ToJavaString(env, PRODUCT_VERSION);
}

// static
void SetServiceWorkerIoThreadClient(
    JNIEnv* env,
    const JavaParamRef<jclass>&,
    const base::android::JavaParamRef<jobject>& io_thread_client,
    const base::android::JavaParamRef<jobject>& browser_context) {
  AwContentsIoThreadClientImpl::SetServiceWorkerIoThreadClient(
      io_thread_client, browser_context);
}


bool RegisterAwContentsStatics(JNIEnv* env) {
  return RegisterNativesImpl(env);
}

}  // namespace android_webview

File: android_webview/java/src/org/chromium/android_webview/AwContentsStatics.java
// Copyright 2014 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package org.chromium.android_webview;

import android.annotation.TargetApi;
import android.content.Context;
import android.webkit.ValueCallback;

import org.chromium.base.ThreadUtils;
import org.chromium.base.annotations.CalledByNative;
import org.chromium.base.annotations.JNINamespace;

import java.lang.reflect.Method;

/**
 * Implementations of various static methods, and also a home for static
 * data structures that are meant to be shared between all webviews.
 */
@JNINamespace("android_webview")
public class AwContentsStatics {

    private static ClientCertLookupTable sClientCertLookupTable;

    private static String sUnreachableWebDataUrl;

    private static boolean sRecordFullDocument;

    private static final String sSafeBrowsingWarmUpHelper =
            "com.android.webview.chromium.SafeBrowsingWarmUpHelper";

    /**
     * Return the client certificate lookup table.
     */
    public static ClientCertLookupTable getClientCertLookupTable() {
        ThreadUtils.assertOnUiThread();
        if (sClientCertLookupTable == null) {
            sClientCertLookupTable = new ClientCertLookupTable();
        }
        return sClientCertLookupTable;
    }

    /**
     * Clear client cert lookup table. Should only be called from UI thread.
     */
    public static void clearClientCertPreferences(Runnable callback) {
        ThreadUtils.assertOnUiThread();
        getClientCertLookupTable().clear();
        nativeClearClientCertPreferences(callback);
    }

    @CalledByNative
    private static void clientCertificatesCleared(Runnable callback) {
        if (callback == null) return;
        callback.run();
    }

    public static String getUnreachableWebDataUrl() {
        // Note that this method may be called from both IO and UI threads,
        // but as it only retrieves a value of a constant from native, even if
        // two calls will be running at the same time, this should not cause
        // any harm.
        if (sUnreachableWebDataUrl == null) {
            sUnreachableWebDataUrl = nativeGetUnreachableWebDataUrl();
        }
        return sUnreachableWebDataUrl;
    }

    public static void setRecordFullDocument(boolean recordFullDocument) {
        sRecordFullDocument = recordFullDocument;
    }

    /* package */ static boolean getRecordFullDocument() {
        return sRecordFullDocument;
    }

    public static void setLegacyCacheRemovalDelayForTest(long timeoutMs) {
        nativeSetLegacyCacheRemovalDelayForTest(timeoutMs);
    }

    public static String getProductVersion() {
        return nativeGetProductVersion();
    }

    public static void setServiceWorkerIoThreadClient(AwContentsIoThreadClient ioThreadClient,
            AwBrowserContext browserContext) {
        nativeSetServiceWorkerIoThreadClient(ioThreadClient, browserContext);
    }

    // Can be called from any thread.
    public static boolean getSafeBrowsingEnabledByManifest() {
        return nativeGetSafeBrowsingEnabledByManifest();
    }

    public static void setSafeBrowsingEnabledByManifest(boolean enable) {
        nativeSetSafeBrowsingEnabledByManifest(enable);
    }

    // TODO(ntfschr): remove this when downstream no longer depends on it
    public static boolean getSafeBrowsingEnabled() {
        return getSafeBrowsingEnabledByManifest();
    }

    // TODO(ntfschr): remove this when downstream no longer depends on it
    public static void setSafeBrowsingEnabled(boolean enable) {
        setSafeBrowsingEnabledByManifest(enable);
    }

    public static void setSafeBrowsingWhiteList(String[] urls) {
        nativeSetSafeBrowsingWhiteList(urls);
    }

    @SuppressWarnings("unchecked")
    @TargetApi(19)
    public static void initSafeBrowsing(Context context, ValueCallback<Boolean> callback) {
        if (callback == null) {
            callback = new ValueCallback<Boolean>() {
                @Override
                public void onReceiveValue(Boolean b) {}
            };
        }

        try {
            Class cls = Class.forName(sSafeBrowsingWarmUpHelper);
            Method m =
                    cls.getDeclaredMethod("warmUpSafeBrowsing", Context.class, ValueCallback.class);
            m.invoke(null, context, callback);
        } catch (ReflectiveOperationException e) {
            callback.onReceiveValue(false);
        }
    }

    @SuppressWarnings("unchecked")
    @TargetApi(19)
    public static void shutdownSafeBrowsing() {
        try {
            Class cls = Class.forName(sSafeBrowsingWarmUpHelper);
            Method m = cls.getDeclaredMethod("coolDownSafeBrowsing");
            m.invoke(null);
        } catch (ReflectiveOperationException e) {
            // This is not an error; it just means this device doesn't have specialized services.
        }
    }

    public static void setCheckClearTextPermitted(boolean permitted) {
        nativeSetCheckClearTextPermitted(permitted);
    }

    /**
     * Return the first substring consisting of the address of a physical location.
     * @see {@link android.webkit.WebView#findAddress(String)}
     *
     * @param addr The string to search for addresses.
     * @return the address, or if no address is found, return null.
     */
    public static String findAddress(String addr) {
        if (addr == null) {
            throw new NullPointerException("addr is null");
        }
        String result = nativeFindAddress(addr);
        return result == null || result.isEmpty() ? null : result;
    }

    //--------------------------------------------------------------------------------------------
    //  Native methods
    //--------------------------------------------------------------------------------------------
    private static native void nativeClearClientCertPreferences(Runnable callback);
    private static native String nativeGetUnreachableWebDataUrl();
    private static native void nativeSetLegacyCacheRemovalDelayForTest(long timeoutMs);
    private static native String nativeGetProductVersion();
    private static native void nativeSetServiceWorkerIoThreadClient(
            AwContentsIoThreadClient ioThreadClient, AwBrowserContext browserContext);
    private static native boolean nativeGetSafeBrowsingEnabledByManifest();
    private static native void nativeSetSafeBrowsingEnabledByManifest(boolean enable);
    private static native void nativeSetSafeBrowsingWhiteList(String[] urls);
    private static native void nativeSetCheckClearTextPermitted(boolean permitted);
    private static native String nativeFindAddress(String addr);
}
